File: src/main/java/com/github/leomillon/uuidgenerator/GenerateToClipboardAction.java
/*
 * The MIT License (MIT)
 *
 *  Copyright (c) 2019 Léo Millon
 *
 *  Permission is hereby granted, free of charge, to any person obtaining a copy
 *  of this software and associated documentation files (the "Software"), to deal
 *  in the Software without restriction, including without limitation the rights
 *  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 *  copies of the Software, and to permit persons to whom the Software is
 *  furnished to do so, subject to the following conditions:
 *
 *  The above copyright notice and this permission notice shall be included in
 *  all copies or substantial portions of the Software.
 *
 *  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 *  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 *  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 *  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 *  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 *  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 *  THE SOFTWARE.
 */
package com.github.leomillon.uuidgenerator;

import java.util.UUID;

import com.intellij.openapi.actionSystem.AnAction;
import com.intellij.openapi.actionSystem.AnActionEvent;
import com.intellij.openapi.actionSystem.CommonDataKeys;
import com.intellij.openapi.ide.CopyPasteManager;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.ui.MessageType;
import com.intellij.openapi.ui.popup.Balloon;
import com.intellij.openapi.ui.popup.JBPopupFactory;
import com.intellij.openapi.wm.StatusBar;
import com.intellij.openapi.wm.WindowManager;
import com.intellij.ui.awt.RelativePoint;
import com.intellij.util.ui.TextTransferable;

/**
 * Action that generate a UUID into clipboard.
 *
 * @author Léo Millon
 */
public class GenerateToClipboardAction extends AnAction {

    @Override
    public void actionPerformed(AnActionEvent anActionEvent) {

        UUID generatedUuid = UUID.randomUUID();

        toClipboard(generatedUuid.toString());

        Project project = anActionEvent.getData(CommonDataKeys.PROJECT);

        if (project != null) {
            displayMessage(generatedUuid.toString() + " copied to clipboard.", project);
        }
    }

    private static void toClipboard(String text) {
        CopyPasteManager.getInstance().setContents(new TextTransferable(text));
    }

    private static void displayMessage(String message, Project project) {
        StatusBar statusBar = WindowManager.getInstance().getStatusBar(project);
        JBPopupFactory.getInstance()
            .createHtmlTextBalloonBuilder(message, MessageType.INFO, null)
            .setFadeoutTime(2000)
            .createBalloon()
            .show(RelativePoint.getCenterOf(statusBar.getComponent()), Balloon.Position.atRight);
    }
}

File: src/main/kotlin/com/github/leomillon/uuidgenerator/GenerateToClipboardAction.kt
package com.github.leomillon.uuidgenerator

import com.github.leomillon.uuidgenerator.DisplayMessageUtils.Companion.displayMessage
import com.intellij.openapi.actionSystem.AnAction
import com.intellij.openapi.actionSystem.AnActionEvent
import com.intellij.openapi.actionSystem.CommonDataKeys
import com.intellij.openapi.ide.CopyPasteManager
import com.intellij.util.ui.TextTransferable
import java.util.*

/**
 * Action that generate a UUID into clipboard.
 *
 * @author Léo Millon
 */
class GenerateToClipboardAction : AnAction() {

    override fun actionPerformed(anActionEvent: AnActionEvent) {

        val generatedUuid = UUID.randomUUID()

        toClipboard(generatedUuid.toString())

        anActionEvent.getData(CommonDataKeys.PROJECT)
            ?.run {
                displayMessage("$generatedUuid copied to clipboard.", this, fadeoutTime = 2000)
            }
    }

    private fun toClipboard(text: String) {
        CopyPasteManager.getInstance().setContents(TextTransferable(text))
    }
}
